### PR TITLE
MM-30540 - display renewal panel in license screen

### PR DESCRIPTION
**Pull request by pablovelezvidal**
#### Summary
MM-30540 - Display renewal panel in License screen
System Console > License screen: show a message indicating that license is up for renewal with button to renew in CWS.

#### Ticket Link
[MM-30540](https://mattermost.atlassian.net/browse/MM-30540)

#### Related Pull Requests

[Depends on](https://github.com/mattermost/mattermost-webapp/pull/7355#issuecomment-767042747)

#### Screenshots
Captura de pantalla 2021-01-23 a las 14 00 34
Captura de pantalla 2021-01-23 a las 14 01 16

### CODE REVIEW

File: components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.scss [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+.RenewLicenseLinkCard__licensedUsersNum {
+    display: inline-block;
+}
+
+.RenewLicenseLinkCard__activeUsersNum {
+    display: inline-block;
+    margin-left: 25px;
+}
```
Comment (marianunez) [inline components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.scss:60]: nit: missing eof line

File: i18n/en.json [hunk 2]
```diff
@@ -1121,6 +1121,9 @@
   "admin.license.keyRemove": "Remove Enterprise License and Downgrade Server",
   "admin.license.noFile": "No file uploaded",
   "admin.license.removing": "Removing License...",
+  "admin.license.subscription.privateCloudCard.description": "Review your numbers below to ensure you renew for theright number of users.",
```
Comment (marianunez) [inline i18n/en.json:1124]: Given that this is not related to Cloud I would change the naming to something like (including the couple below: 

```suggestion
  "admin.license.renewalCard.description": "Review your numbers below to ensure you renew for the right number of users.",
```
(There was also a space missing)

File: components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import React from 'react';
+
+import {FormattedMessage} from 'react-intl';
+
+import moment from 'moment';
+
+import RenewalLink from 'components/announcement_bar/renewal_link/';
+
+import alertIcon from 'images/icons/round-blue-info-icon.svg';
+import warningIcon from 'images/icons/warning-red-icon.svg';
```
Comment (marianunez) [inline components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx:12]: I believe we may already have these icons in the [compass icons font](https://github.com/mattermost/mattermost-webapp/blob/master/sass/base/_compass-icons.scss). 

I believe it's `alert-circle-outline` for the alert and ` alert-outline` for the warning. You can see an example on how to use them [here](https://github.com/mattermost/mattermost-webapp/blob/master/components/alert_banner.tsx#L27)

File: components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+import './renew_license_link_card.scss';
+
+export interface RenewLicenseLinkCardProps {
+    license: any;
+    isLicenseExpired: boolean;
+    totalUsers: number;
+}
+
+const RenewLicenseLinkCard: React.FC<RenewLicenseLinkCardProps> = ({license, totalUsers, isLicenseExpired}: RenewLicenseLinkCardProps) => {
```
Comment (marianunez) [inline components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx:24]: Possible name to simplify here, `Link` may not be adding any value here

```suggestion
const RenewLicenseCard: React.FC<RenewLicenseCardProps> = ({license, totalUsers, isLicenseExpired}: RenewLicenseCardProps) => {
```

File: components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    let titleClass = 'RenewLicenseLinkCard__text-title';
+    let icon = alertIcon;
+    let idTitleText = 'admin.license.license_expiring';
+    let defaultTitleText = 'License expires in {days} days on {date, date, long}.';
```
Comment (marianunez) [inline components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx:28]: Do we need to add this to the translation file?

File: components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                <div className='RenewLicenseLinkCard__text-description bolder'>
+                    <FormattedMessage
+                        id='admin.license.subscription.privateCloudCard.description'
+                        defaultMessage='Renew your Enterprise license through the Customer Portal to avoid any disruption.'
+                    />
+                </div>
+                <div className='RenewLicenseLinkCard__text-description'>
+                    <FormattedMessage
+                        id='admin.license.subscription.privateCloudCard.description'
+                        defaultMessage='Review your numbers below to ensure you renew for theright number of users.'
```
Comment (marianunez) [inline components/admin_console/license_settings/renew_license_link_card/renew_license_link_card.tsx:64]:
```suggestion
                        defaultMessage='Review your numbers below to ensure you renew for the right number of users.'
```

File: components/admin_console/license_settings/license_settings.jsx [hunk 7]
```diff
@@ -605,5 +611,18 @@ export default class LicenseSettings extends React.PureComponent {
...
         );
     }
+
+    renewLicenseCard = () => {
+        if (isLicenseExpired(this.props.license) || isLicenseExpiring(this.props.license)) {
+            return (
+                <RenewLinkCard
+                    license={this.props.license}
+                    isLicenseExpired={isLicenseExpired(this.props.license)}
+                    totalUsers={Math.max(this.props.stats.TOTAL_USERS, 30)}
```
Comment (marianunez) [inline components/admin_console/license_settings/license_settings.jsx:621]: Why max here to 30 users? We should show them the total users they are currently have in their system even if less than 30

**Review by marianunez — CHANGES_REQUESTED** [review]
Looks great @pablovelezvidal! Some minor comments below

**Review by marianunez — CHANGES_REQUESTED** [review]
Looks great @pablovelezvidal! Some minor comments below

File: components/admin_console/license_settings/license_settings.jsx [hunk 8]
```diff
@@ -405,6 +408,9 @@ export default class LicenseSettings extends React.PureComponent {
                             className='form-horizontal'
                             role='form'
                         >
+                            <div className='license-expir-ing-ed-notice'>
```
Comment (marianunez) [inline components/admin_console/license_settings/license_settings.jsx:411]: Where do we define this style?

**Review by marianunez — APPROVED** [review]
Great work @pablovelezvidal! Thanks!

File: components/admin_console/license_settings/license_settings.jsx (hunk 8 above)
Comment (pablovelezvidal) [reply components/admin_console/license_settings/license_settings.jsx:411]: this is a left-behind code from the initial design that needs to be removed.

**Review by annelieseklein — COMMENTED** [review]
looks good - just one question. it says check the numbers to make sure you renew for the right number of seats. If it isn't the right number, where should they go?

**Comment by marianunez** [comment]
> looks good - just one question. it says check the numbers to make sure you renew for the right number of seats. If it isn't the right number, where should they go?

Good question, there was no indication of any other CTA. See [Figma](https://www.figma.com/file/PJN5OSUNBkDSjR4J9OyC9Q/Customer-Renewal?node-id=2044%3A36354) for reference

**Comment by marianunez** [comment]
Adding @hahmadia for Enterprise team visibility to System Console changes

**Review by annelieseklein — APPROVED** [review]

File: components/admin_console/license_settings/renew_license_card/renew_license_card.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                <div className='RenewLicenseCard__activeUsersNum'>
+                    <FormattedMarkdownMessage
+                        id='admin.license.renewalCard.usersNumbers'
+                        defaultMessage='**Active Users:** {activeUsersNum}'
+                        values={{
+                            activeUsersNum: totalUsers,
+                        }}
+                    />
+                </div>
+                <RenewalLink/>
```
Comment (AGMETEOR) [inline components/admin_console/license_settings/renew_license_card/renew_license_card.tsx:79]: Hey @pablovelezvidal , thanks a lot for the good work. I just found some issues with this component here ```<RenewalLink/>```. 
> With a cloud license, we know that ```api/v4/license/renewal``` will not return a renewal link. This means in that component we don't set any link in the state. That leads to a somewhat misleading message on the modal. See image
Screenshot 2021-01-27 at 13 46 29

> Also, because the ```useEffect``` hook in that component runs once, changing the licence to like an E10 or E20 and then clicking the RenewalLink button will still show that misleading message of no internet connection

**Review by AGMETEOR — CHANGES_REQUESTED** [review]

File: components/admin_console/license_settings/renew_license_card/renew_license_card.tsx (hunk 9 above)
Comment (pablovelezvidal) [reply components/admin_console/license_settings/renew_license_card/renew_license_card.tsx:87]: thanks @AGMETEOR for your comments! I will check for more details behind the business logic of these use cases when it is a cloud license and when you change the type of license.

Comment (pablovelezvidal) [reply components/admin_console/license_settings/renew_license_card/renew_license_card.tsx:87]: Hi @AGMETEOR , thanks again for the review! so, no Cloud customer is supposed to see the Edition and License screen, meaning that the scenario about receiving a bad request for the renewal link is only happening to us in development for a special configuration that we use to see this page. Also, we want to keep it that way for the hypothetical scenario where our Cloud License is expiring and this way we can easily see that instead of it to fail silently. The other scenario mentioned is not a problem, because every time that you remove/add a license, the entire component and child tree is re-rendered, see attached gif.

licenseCard

Comment (AGMETEOR) [reply components/admin_console/license_settings/renew_license_card/renew_license_card.tsx:86]: Alright. Thanks a lot @pablovelezvidal

**Review by AGMETEOR — APPROVED** [review]

File: components/announcement_bar/renewal_link/renewal_link.tsx [hunk 10]
```diff
@@ -24,8 +24,12 @@ const RenewalLink: React.FC<RenewalLinkProps> = (props: RenewalLinkProps) => {
...
     useEffect(() => {
         Client4.getRenewalLink().then(({renewal_link: renewalLinkParam}) => {
-            if (renewalLinkParam && (/^http[s]?:\/\//).test(renewalLinkParam)) {
-                setRenewalLink(renewalLinkParam);
+            try {
+                if (renewalLinkParam && (/^http[s]?:\/\//).test(renewalLinkParam)) {
+                    setRenewalLink(renewalLinkParam);
+                }
+            } catch (error) {
+                console.error('No link returned', error); // eslint-disable-line no-console
```
Comment (hahmadia) [inline components/announcement_bar/renewal_link/renewal_link.tsx:32]: We wouldn't want to show this to customer right? Would console logging it be enough?

**Review by hahmadia — APPROVED** [review]
Thanks Maria for @ me on this PR for visibility.
Code looks good with one questions.

File: components/announcement_bar/renewal_link/renewal_link.tsx (hunk 10 above)
Comment (pablovelezvidal) [reply components/announcement_bar/renewal_link/renewal_link.tsx:32]: yes, we handle the error gracefully,  we catch it and we write it to the console.

**Comment by marianunez** [comment]
Steve is OOO so I've tested both cases (license to expire and license expired) in the test PR for the `QA Review` step. This will be cherry-picked to 5.32 and will also be part of RC testing. Setting `QA Review Done` tag.
